File: examples/backtest/tardis_option_chain.py
#!/usr/bin/env python3
# -------------------------------------------------------------------------------------------------
#  Copyright (C) 2015-2026 Nautech Systems Pty Ltd. All rights reserved.
#  https://nautechsystems.io
#
#  Licensed under the GNU Lesser General Public License Version 3.0 (the "License");
#  You may not use this file except in compliance with the License.
#  You may obtain a copy of the License at https://www.gnu.org/licenses/lgpl-3.0.en.html
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
# -------------------------------------------------------------------------------------------------
"""
Example: option-chain backtest from a Tardis-backed catalog.

The catalog must already contain option instruments plus per-instrument QuoteTick
and OptionGreeks data, such as data written by the Tardis Machine replay pipeline.

Run with a built v2 extension:
    python examples/backtest/tardis_option_chain.py --catalog-path /path/to/catalog
"""

from __future__ import annotations

import argparse
from dataclasses import dataclass
from decimal import Decimal
from pathlib import Path
from typing import Any
from typing import Self

from nautilus_trader.backtest import BacktestDataConfig  # type: ignore[attr-defined]
from nautilus_trader.backtest import BacktestEngineConfig  # type: ignore[attr-defined]
from nautilus_trader.backtest import BacktestNode  # type: ignore[attr-defined]
from nautilus_trader.backtest import BacktestRunConfig  # type: ignore[attr-defined]
from nautilus_trader.backtest import BacktestVenueConfig  # type: ignore[attr-defined]
from nautilus_trader.core import UUID4
from nautilus_trader.execution import CappedOptionFeeModel  # type: ignore[attr-defined]
from nautilus_trader.execution import TieredNotionalOptionFeeModel  # type: ignore[attr-defined]
from nautilus_trader.model import AccountType  # type: ignore[attr-defined]
from nautilus_trader.model import BookType  # type: ignore[attr-defined]
from nautilus_trader.model import ClientOrderId
from nautilus_trader.model import ContingencyType  # type: ignore[attr-defined]
from nautilus_trader.model import InstrumentId
from nautilus_trader.model import LimitOrder  # type: ignore[attr-defined]
from nautilus_trader.model import MarketOrder  # type: ignore[attr-defined]
from nautilus_trader.model import OmsType  # type: ignore[attr-defined]
from nautilus_trader.model import OptionChainSlice  # type: ignore[attr-defined]
from nautilus_trader.model import OptionSeriesId  # type: ignore[attr-defined]
from nautilus_trader.model import OrderSide  # type: ignore[attr-defined]
from nautilus_trader.model import Price
from nautilus_trader.model import Quantity
from nautilus_trader.model import StrikeRange  # type: ignore[attr-defined]
from nautilus_trader.model import TimeInForce  # type: ignore[attr-defined]
from nautilus_trader.model import TraderId
from nautilus_trader.persistence import ParquetDataCatalog  # type: ignore[attr-defined]
from nautilus_trader.trading import ImportableStrategyConfig  # type: ignore[attr-defined]
from nautilus_trader.trading import Strategy
from nautilus_trader.trading import StrategyConfig  # type: ignore[attr-defined]


VENUE = "DERIBIT"


@dataclass(frozen=True)
class OptionMetadata:
    instrument_id: InstrumentId
    underlying: str
    settlement_currency: str
    expiration_ns: int
    strike: Price


@dataclass(frozen=True)
class SeriesSelection:
    series_id: OptionSeriesId
    instrument_ids: list[InstrumentId]
    strikes: list[Price]
    settlement_currency: str


@dataclass(frozen=True)
class SelectedOption:
    instrument_id: InstrumentId
    strike: Price
    quote: Any
    delta: float | None


class OptionChainBacktestConfig(StrategyConfig):
    _CUSTOM_FIELDS = (
        "series_id",
        "selection_mode",
        "target_delta",
        "delta_tolerance",
        "target_strike",
        "trade_size",
        "snapshot_interval_ms",
    )

    def __new__(cls, *args: Any, **kwargs: Any) -> Self:
        for key in cls._CUSTOM_FIELDS:
            kwargs.pop(key, None)
        return super().__new__(cls, *args, **kwargs)

    def __init__(
        self,
        series_id: str,
        selection_mode: str = "delta",
        target_delta: float = 0.25,
        delta_tolerance: float = 0.05,
        target_strike: str | None = None,
        trade_size: str = "1",
        snapshot_interval_ms: int = 1_000,
        **kwargs: Any,
    ) -> None:
        super().__init__()
        self.series_id = series_id
        self.selection_mode = selection_mode
        self.target_delta = target_delta
        self.delta_tolerance = delta_tolerance
        self.target_strike = target_strike
        self.trade_size = trade_size
        self.snapshot_interval_ms = snapshot_interval_ms


class OptionChainBacktest(Strategy):
    def __init__(self, config: OptionChainBacktestConfig) -> None:
        super().__init__(config)
        self._series_id = OptionSeriesId.from_str(config.series_id)
        self._selection_mode = config.selection_mode
        self._target_delta = config.target_delta
        self._delta_tolerance = config.delta_tolerance
        self._target_strike = Price.from_str(config.target_strike) if config.target_strike else None
        self._trade_size = Quantity.from_str(config.trade_size)
        self._snapshot_interval_ms = config.snapshot_interval_ms
        self._orders_submitted = False

    def on_start(self) -> None:
        if self._selection_mode == "strike":
            if self._target_strike is None:
                raise ValueError("target_strike is required when selection_mode is 'strike'")
            strike_range = StrikeRange.fixed([self._target_strike])
        else:
            strike_range = StrikeRange.delta(self._target_delta, self._delta_tolerance)

        self.subscribe_option_chain(
            self._series_id,
            strike_range=strike_range,
            snapshot_interval_ms=self._snapshot_interval_ms,
        )

    def on_option_chain(self, slice: OptionChainSlice) -> None:
        self.log.info(
            f"OPTION_CHAIN | {slice.series_id} | atm={slice.atm_strike} | "
            f"calls={slice.call_count()} puts={slice.put_count()} strikes={slice.strike_count()}",
        )

        if self._orders_submitted:
            return

        selected = self._select_contract(slice)
        if selected is None:
            return

        self.log.info(
            f"Selected option {selected.instrument_id} at strike {selected.strike} "
            f"with delta {selected.delta}",
        )
        self.submit_order(self._maker_order(selected))
        self.submit_order(self._taker_order(selected))
        self._orders_submitted = True

    def on_stop(self) -> None:
        self.unsubscribe_option_chain(self._series_id)

    def _select_contract(self, slice: OptionChainSlice) -> SelectedOption | None:
        if self._selection_mode == "strike":
            return self._select_by_strike(slice)
        return self._select_by_delta(slice) or self._first_quoted_contract(slice)

    def _select_by_delta(self, slice: OptionChainSlice) -> SelectedOption | None:
        best: tuple[float, SelectedOption] | None = None

        for strike in slice.strikes():
            for data in (slice.get_call(strike), slice.get_put(strike)):
                if data is None or data.greeks is None:
                    continue
                distance = abs(abs(data.greeks.delta) - self._target_delta)
                if distance > self._delta_tolerance:
                    continue
                selected = SelectedOption(
                    instrument_id=data.quote.instrument_id,
                    strike=strike,
                    quote=data.quote,
                    delta=data.greeks.delta,
                )

                if best is None or distance < best[0]:
                    best = (distance, selected)

        return best[1] if best is not None else None

    def _select_by_strike(self, slice: OptionChainSlice) -> SelectedOption | None:
        if self._target_strike is None:
            return None

        data = slice.get_call(self._target_strike) or slice.get_put(self._target_strike)
        if data is None:
            return None
        return SelectedOption(
            instrument_id=data.quote.instrument_id,
            strike=self._target_strike,
            quote=data.quote,
            delta=data.greeks.delta if data.greeks is not None else None,
        )

    def _first_quoted_contract(self, slice: OptionChainSlice) -> SelectedOption | None:
        for strike in slice.strikes():
            data = slice.get_call(strike) or slice.get_put(strike)
            if data is not None:
                return SelectedOption(
                    instrument_id=data.quote.instrument_id,
                    strike=strike,
                    quote=data.quote,
                    delta=data.greeks.delta if data.greeks is not None else None,
                )
        return None

    def _maker_order(self, selected: SelectedOption) -> LimitOrder:
        return LimitOrder(
            trader_id=self.trader_id,
            strategy_id=self.strategy_id,
            instrument_id=selected.instrument_id,
            client_order_id=self._client_order_id("M"),
            order_side=OrderSide.BUY,
            quantity=self._trade_size,
            price=selected.quote.bid_price,
            time_in_force=TimeInForce.GTC,
            post_only=True,
            reduce_only=False,
            quote_quantity=False,
            init_id=UUID4(),
            ts_init=self.clock.timestamp_ns(),
            contingency_type=ContingencyType.NO_CONTINGENCY,
        )

    def _taker_order(self, selected: SelectedOption) -> MarketOrder:
        return MarketOrder(
            trader_id=self.trader_id,
            strategy_id=self.strategy_id,
            instrument_id=selected.instrument_id,
            client_order_id=self._client_order_id("T"),
            order_side=OrderSide.BUY,
            quantity=self._trade_size,
            init_id=UUID4(),
            ts_init=self.clock.timestamp_ns(),
            time_in_force=TimeInForce.GTC,
            reduce_only=False,
            quote_quantity=False,
            contingency_type=ContingencyType.NO_CONTINGENCY,
        )

    def _client_order_id(self, label: str) -> ClientOrderId:
        return ClientOrderId(f"{self.strategy_id}-{label}-{UUID4()}")


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("--catalog-path", type=Path, default=Path("./catalog"))
    parser.add_argument("--underlying", default="BTC")
    parser.add_argument("--selection", choices=("delta", "strike"), default="delta")
    parser.add_argument("--target-delta", type=float, default=0.25)
    parser.add_argument("--delta-tolerance", type=float, default=0.05)
    parser.add_argument("--target-strike")
    parser.add_argument("--fee-model", choices=("capped", "tiered"), default="capped")
    parser.add_argument("--chunk-size", type=int, default=10_000)
    return parser.parse_args()


def main() -> None:
    args = parse_args()
    options = load_option_metadata(args.catalog_path, args.underlying)
    selection = nearest_series(options)
    target_strike = args.target_strike

    if args.selection == "strike" and target_strike is None:
        target_strike = str(median_strike(selection.strikes))

    venue_config = BacktestVenueConfig(
        name=VENUE,
        oms_type=OmsType.NETTING,
        account_type=AccountType.MARGIN,
        book_type=BookType.L1_MBP,
        starting_balances=[starting_balance(selection.settlement_currency)],
        fee_model=option_fee_model(args.fee_model),
    )
    data = [
        BacktestDataConfig(
            data_type="QuoteTick",
            catalog_path=str(args.catalog_path),
            instrument_ids=selection.instrument_ids,
        ),
        BacktestDataConfig(
            data_type="OptionGreeks",
            catalog_path=str(args.catalog_path),
            instrument_ids=selection.instrument_ids,
        ),
    ]
    run_config = BacktestRunConfig(
        id="tardis-option-chain",
        engine=BacktestEngineConfig(
            trader_id=TraderId("OPTION-CHAIN-001"),
            run_analysis=False,
        ),
        venues=[venue_config],
        data=data,
        chunk_size=args.chunk_size,
    )

    module_name = Path(__file__).stem
    strategy_config = ImportableStrategyConfig(
        strategy_path=f"{module_name}:OptionChainBacktest",
        config_path=f"{module_name}:OptionChainBacktestConfig",
        config={
            "series_id": str(selection.series_id),
            "selection_mode": args.selection,
            "target_delta": args.target_delta,
            "delta_tolerance": args.delta_tolerance,
            "target_strike": target_strike,
            "trade_size": "1",
            "snapshot_interval_ms": 1_000,
        },
    )

    node = BacktestNode([run_config])
    node.build()
    node.add_strategy_from_config(run_config.id, strategy_config)
    node.run()
    node.dispose()


def load_option_metadata(catalog_path: Path, underlying: str) -> list[OptionMetadata]:
    catalog = ParquetDataCatalog(str(catalog_path))
    options = [
        metadata
        for instrument in catalog.instruments()
        if (metadata := option_metadata(instrument)) is not None
        and metadata.instrument_id.venue.value == VENUE
        and metadata.underlying.upper() == underlying.upper()
    ]

    if not options:
        raise RuntimeError(
            f"No {underlying} option instruments found for {VENUE} in {catalog_path}",
        )
    return options


def option_metadata(instrument: Any) -> OptionMetadata | None:
    if instrument.type_name == "CryptoOption":
        return OptionMetadata(
            instrument_id=instrument.id,
            underlying=instrument.underlying.code,
            settlement_currency=instrument.settlement_currency.code,
            expiration_ns=instrument.expiration_ns,
            strike=instrument.strike_price,
        )
    if instrument.type_name == "OptionContract":
        return OptionMetadata(
            instrument_id=instrument.id,
            underlying=instrument.underlying,
            settlement_currency=instrument.currency.code,
            expiration_ns=instrument.expiration_ns,
            strike=instrument.strike_price,
        )
    return None


def nearest_series(options: list[OptionMetadata]) -> SeriesSelection:
    expiration_ns = min(metadata.expiration_ns for metadata in options)
    same_expiry = [metadata for metadata in options if metadata.expiration_ns == expiration_ns]
    settlement_currency = next(
        (
            metadata.settlement_currency
            for metadata in same_expiry
            if metadata.settlement_currency == metadata.underlying
        ),
        same_expiry[0].settlement_currency,
    )
    instrument_ids = [
        metadata.instrument_id
        for metadata in same_expiry
        if metadata.settlement_currency == settlement_currency
    ]
    strikes = sorted(
        {
            metadata.strike
            for metadata in same_expiry
            if metadata.settlement_currency == settlement_currency
        },
    )
    series_id = OptionSeriesId(
        VENUE,
        same_expiry[0].underlying,
        settlement_currency,
        expiration_ns,
    )
    return SeriesSelection(series_id, instrument_ids, strikes, settlement_currency)


def median_strike(strikes: list[Price]) -> Price:
    return strikes[len(strikes) // 2]


def option_fee_model(name: str) -> CappedOptionFeeModel | TieredNotionalOptionFeeModel:
    if name == "tiered":
        return TieredNotionalOptionFeeModel(
            maker_rate=Decimal("0.0002"),
            taker_rate=Decimal("0.0005"),
        )
    return CappedOptionFeeModel(
        maker_rate=Decimal("0.0003"),
        taker_rate=Decimal("0.0003"),
    )


def starting_balance(settlement_currency: str) -> str:
    if settlement_currency in {"BTC", "ETH"}:
        return f"10 {settlement_currency}"
    return f"1000000 {settlement_currency}"


if __name__ == "__main__":
    main()

File: crates/backtest/examples/tardis_option_chain.rs
// -------------------------------------------------------------------------------------------------
//  Copyright (C) 2015-2026 Nautech Systems Pty Ltd. All rights reserved.
//  https://nautechsystems.io
//
//  Licensed under the GNU Lesser General Public License Version 3.0 (the "License");
//  You may not use this file except in compliance with the License.
//  You may obtain a copy of the License at https://www.gnu.org/licenses/lgpl-3.0.en.html
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// -------------------------------------------------------------------------------------------------

//! Example: option-chain backtest from a Tardis-backed catalog.
//!
//! The catalog must already contain option instruments plus per-instrument
//! `QuoteTick` and `OptionGreeks` data, such as data written by the Tardis
//! Machine replay pipeline. Edit `CATALOG_PATH` below to point at it.
//!
//! Edit the constants below to change the catalog, underlying, contract
//! selection, and fee model.
//!
//! Run with:
//! `cargo run -p nautilus-backtest --features examples,streaming --example tardis-option-chain`

use std::{fmt::Debug, path::Path, str::FromStr};

use nautilus_backtest::{
    config::{BacktestDataConfig, BacktestRunConfig, BacktestVenueConfig, NautilusDataType},
    node::BacktestNode,
};
use nautilus_common::actor::DataActor;
use nautilus_core::UnixNanos;
use nautilus_execution::models::fee::{
    CappedOptionFeeModel, FeeModelAny, TieredNotionalOptionFeeModel,
};
use nautilus_model::{
    data::{
        QuoteTick,
        option_chain::{OptionChainSlice, StrikeRange},
    },
    enums::{AccountType, BookType, OmsType, OrderSide, TimeInForce},
    identifiers::{InstrumentId, OptionSeriesId, StrategyId, Venue},
    instruments::InstrumentAny,
    types::{Price, Quantity},
};
use nautilus_persistence::backend::catalog::ParquetDataCatalog;
use nautilus_trading::{Strategy, StrategyConfig, StrategyCore, nautilus_strategy};
use rust_decimal::Decimal;
use ustr::Ustr;

const VENUE: &str = "DERIBIT";
const CATALOG_PATH: &str = "./catalog";
const UNDERLYING: &str = "BTC";
const SELECTION: &str = "delta"; // "delta" or "strike"
const TARGET_DELTA: f64 = 0.25;
const DELTA_TOLERANCE: f64 = 0.05;
const TARGET_STRIKE: Option<&str> = None; // None selects the median strike
const FEE_MODEL: &str = "capped"; // "capped" or "tiered"

#[derive(Debug, Clone)]
struct OptionMetadata {
    instrument_id: InstrumentId,
    underlying: Ustr,
    settlement_currency: Ustr,
    expiration_ns: u64,
    strike: Price,
}

#[derive(Debug, Clone, Copy)]
enum SelectionMode {
    Delta { target: f64, tolerance: f64 },
    Strike { strike: Price },
}

#[derive(Debug, Clone, Copy)]
struct SelectedOption {
    instrument_id: InstrumentId,
    strike: Price,
    quote: QuoteTick,
    delta: Option<f64>,
}

#[derive(Debug)]
struct OptionChainBacktest {
    core: StrategyCore,
    series_id: OptionSeriesId,
    strike_range: StrikeRange,
    selection_mode: SelectionMode,
    trade_size: Quantity,
    orders_submitted: bool,
}

impl OptionChainBacktest {
    fn new(
        series_id: OptionSeriesId,
        strike_range: StrikeRange,
        selection_mode: SelectionMode,
        trade_size: Quantity,
    ) -> Self {
        let config = StrategyConfig {
            strategy_id: Some(StrategyId::from("OPTION-CHAIN-001")),
            order_id_tag: Some("001".to_string()),
            ..Default::default()
        };
        Self {
            core: StrategyCore::new(config),
            series_id,
            strike_range,
            selection_mode,
            trade_size,
            orders_submitted: false,
        }
    }

    fn select_contract(&self, slice: &OptionChainSlice) -> Option<SelectedOption> {
        match self.selection_mode {
            SelectionMode::Delta { target, tolerance } => {
                self.select_by_delta(slice, target, tolerance)
            }
            SelectionMode::Strike { strike } => self.select_by_strike(slice, strike),
        }
    }

    fn select_by_delta(
        &self,
        slice: &OptionChainSlice,
        target: f64,
        tolerance: f64,
    ) -> Option<SelectedOption> {
        let mut best: Option<(f64, SelectedOption)> = None;

        for (strike, data) in slice.calls.iter().chain(slice.puts.iter()) {
            let Some(greeks) = data.greeks.as_ref() else {
                continue;
            };
            let distance = (greeks.delta.abs() - target).abs();
            if distance > tolerance {
                continue;
            }
            let selected = SelectedOption {
                instrument_id: data.quote.instrument_id,
                strike: *strike,
                quote: data.quote,
                delta: Some(greeks.delta),
            };

            if best
                .as_ref()
                .is_none_or(|(best_distance, _)| distance < *best_distance)
            {
                best = Some((distance, selected));
            }
        }

        best.map(|(_, selected)| selected)
            .or_else(|| first_quoted_contract(slice))
    }

    fn select_by_strike(&self, slice: &OptionChainSlice, strike: Price) -> Option<SelectedOption> {
        slice
            .get_call(&strike)
            .or_else(|| slice.get_put(&strike))
            .map(|data| SelectedOption {
                instrument_id: data.quote.instrument_id,
                strike,
                quote: data.quote,
                delta: data.greeks.as_ref().map(|greeks| greeks.delta),
            })
    }

    fn submit_example_orders(&mut self, selected: SelectedOption) -> anyhow::Result<()> {
        log::info!(
            "Selected option {} at strike {} with delta {:?}",
            selected.instrument_id,
            selected.strike,
            selected.delta,
        );

        let trade_size = self.trade_size;
        let maker_order = self.order().limit(
            selected.instrument_id,
            OrderSide::Buy,
            trade_size,
            selected.quote.bid_price,
            Some(TimeInForce::Gtc),
            None,
            Some(true),
            None,
            None,
            None,
            None,
            None,
            None,
            None,
            None,
            None,
        );
        self.submit_order(maker_order, None, None, None)?;

        let taker_order = self.order().market(
            selected.instrument_id,
            OrderSide::Buy,
            trade_size,
            Some(TimeInForce::Gtc),
            None,
            None,
            None,
            None,
            None,
            None,
        );
        self.submit_order(taker_order, None, None, None)?;

        self.orders_submitted = true;
        Ok(())
    }
}

nautilus_strategy!(OptionChainBacktest);

impl DataActor for OptionChainBacktest {
    fn on_start(&mut self) -> anyhow::Result<()> {
        self.subscribe_option_chain(
            self.series_id,
            self.strike_range.clone(),
            Some(1_000),
            None,
            None,
        );
        Ok(())
    }

    fn on_option_chain(&mut self, slice: &OptionChainSlice) -> anyhow::Result<()> {
        log::info!(
            "OPTION_CHAIN | {} | atm={:?} | calls={} puts={} strikes={}",
            slice.series_id,
            slice.atm_strike,
            slice.call_count(),
            slice.put_count(),
            slice.strike_count(),
        );

        if self.orders_submitted {
            return Ok(());
        }

        if let Some(selected) = self.select_contract(slice) {
            self.submit_example_orders(selected)?;
        }

        Ok(())
    }

    fn on_stop(&mut self) -> anyhow::Result<()> {
        self.unsubscribe_option_chain(self.series_id, None);
        Ok(())
    }
}

fn main() -> anyhow::Result<()> {
    nautilus_common::logging::ensure_logging_initialized();

    let catalog_path = CATALOG_PATH.to_string();
    let underlying = UNDERLYING.to_string();
    let venue = Venue::new(VENUE);

    let options = load_option_metadata(&catalog_path, venue, &underlying)?;
    let (series_id, instrument_ids, series_options) = nearest_series(venue, &options)?;
    let selection_mode = selection_mode(&series_options)?;
    let strike_range = strike_range(selection_mode);

    let venue_config = BacktestVenueConfig::builder()
        .name(Ustr::from(VENUE))
        .oms_type(OmsType::Netting)
        .account_type(AccountType::Margin)
        .book_type(BookType::L1_MBP)
        .starting_balances(vec![starting_balance(series_id.settlement_currency)])
        .fee_model(option_fee_model()?)
        .build();

    let quote_data = BacktestDataConfig::builder()
        .data_type(NautilusDataType::QuoteTick)
        .catalog_path(catalog_path.clone())
        .instrument_ids(instrument_ids.clone())
        .build();
    let greeks_data = BacktestDataConfig::builder()
        .data_type(NautilusDataType::OptionGreeks)
        .catalog_path(catalog_path)
        .instrument_ids(instrument_ids)
        .build();

    let run_config = BacktestRunConfig::builder()
        .id("tardis-option-chain".to_string())
        .venues(vec![venue_config])
        .data(vec![quote_data, greeks_data])
        .chunk_size(10_000)
        .build();
    let config_id = run_config.id().to_string();

    let mut node = BacktestNode::new(vec![run_config])?;
    node.build()?;

    let engine = node
        .get_engine_mut(&config_id)
        .ok_or_else(|| anyhow::anyhow!("Backtest engine not built for run {config_id}"))?;
    engine.add_strategy(OptionChainBacktest::new(
        series_id,
        strike_range,
        selection_mode,
        Quantity::from("1"),
    ))?;

    node.run()?;

    Ok(())
}

fn load_option_metadata(
    catalog_path: &str,
    venue: Venue,
    underlying: &str,
) -> anyhow::Result<Vec<OptionMetadata>> {
    let catalog = ParquetDataCatalog::new(Path::new(catalog_path), None, None, None, None);
    let instruments = catalog.query_instruments(None)?;

    let options: Vec<OptionMetadata> = instruments
        .iter()
        .filter_map(option_metadata)
        .filter(|metadata| {
            metadata.instrument_id.venue == venue
                && metadata
                    .underlying
                    .as_str()
                    .eq_ignore_ascii_case(underlying)
        })
        .collect();

    if options.is_empty() {
        anyhow::bail!(
            "No {underlying} option instruments found for {venue} in catalog {catalog_path}"
        );
    }
    Ok(options)
}

fn nearest_series(
    venue: Venue,
    options: &[OptionMetadata],
) -> anyhow::Result<(OptionSeriesId, Vec<InstrumentId>, Vec<OptionMetadata>)> {
    let expiration_ns = options
        .iter()
        .map(|metadata| metadata.expiration_ns)
        .min()
        .ok_or_else(|| anyhow::anyhow!("No option expirations found"))?;
    let first = options
        .iter()
        .find(|metadata| metadata.expiration_ns == expiration_ns)
        .ok_or_else(|| anyhow::anyhow!("No options found for expiration {expiration_ns}"))?;
    let settlement_currency = options
        .iter()
        .find(|metadata| {
            metadata.expiration_ns == expiration_ns
                && metadata.settlement_currency == metadata.underlying
        })
        .map_or(first.settlement_currency, |metadata| {
            metadata.settlement_currency
        });
    let series_options: Vec<OptionMetadata> = options
        .iter()
        .filter(|metadata| {
            metadata.expiration_ns == expiration_ns
                && metadata.settlement_currency == settlement_currency
        })
        .cloned()
        .collect();
    let instrument_ids = series_options
        .iter()
        .map(|metadata| metadata.instrument_id)
        .collect();
    let series_id = OptionSeriesId::new(
        venue,
        first.underlying,
        settlement_currency,
        UnixNanos::from(expiration_ns),
    );

    Ok((series_id, instrument_ids, series_options))
}

fn option_metadata(instrument: &InstrumentAny) -> Option<OptionMetadata> {
    match instrument {
        InstrumentAny::CryptoOption(option) => Some(OptionMetadata {
            instrument_id: option.id,
            underlying: option.underlying.code,
            settlement_currency: option.settlement_currency.code,
            expiration_ns: option.expiration_ns.as_u64(),
            strike: option.strike_price,
        }),
        InstrumentAny::OptionContract(option) => Some(OptionMetadata {
            instrument_id: option.id,
            underlying: option.underlying,
            settlement_currency: option.currency.code,
            expiration_ns: option.expiration_ns.as_u64(),
            strike: option.strike_price,
        }),
        _ => None,
    }
}

fn selection_mode(options: &[OptionMetadata]) -> anyhow::Result<SelectionMode> {
    match SELECTION {
        "delta" => Ok(SelectionMode::Delta {
            target: TARGET_DELTA,
            tolerance: DELTA_TOLERANCE,
        }),
        "strike" => {
            let strike = match TARGET_STRIKE {
                Some(raw) => Price::from(raw),
                None => median_strike(options)?,
            };
            Ok(SelectionMode::Strike { strike })
        }
        other => {
            anyhow::bail!("Invalid SELECTION '{other}', expected 'delta' or 'strike'")
        }
    }
}

fn strike_range(selection_mode: SelectionMode) -> StrikeRange {
    match selection_mode {
        SelectionMode::Delta { target, tolerance } => StrikeRange::Delta { target, tolerance },
        SelectionMode::Strike { strike } => StrikeRange::Fixed(vec![strike]),
    }
}

fn option_fee_model() -> anyhow::Result<FeeModelAny> {
    match FEE_MODEL {
        "capped" => Ok(FeeModelAny::CappedOption(CappedOptionFeeModel::new(
            Some(parse_decimal("0.0003")?),
            Some(parse_decimal("0.0003")?),
            None,
        )?)),
        "tiered" => Ok(FeeModelAny::TieredNotionalOption(
            TieredNotionalOptionFeeModel::new(
                Some(parse_decimal("0.0002")?),
                Some(parse_decimal("0.0005")?),
            )?,
        )),
        other => {
            anyhow::bail!("Invalid FEE_MODEL '{other}', expected 'capped' or 'tiered'")
        }
    }
}

fn parse_decimal(raw: &str) -> anyhow::Result<Decimal> {
    Decimal::from_str(raw).map_err(Into::into)
}

fn median_strike(options: &[OptionMetadata]) -> anyhow::Result<Price> {
    let mut strikes: Vec<Price> = options.iter().map(|metadata| metadata.strike).collect();
    strikes.sort();
    strikes.dedup();
    strikes
        .get(strikes.len() / 2)
        .copied()
        .ok_or_else(|| anyhow::anyhow!("No strikes found in option metadata"))
}

fn first_quoted_contract(slice: &OptionChainSlice) -> Option<SelectedOption> {
    slice
        .calls
        .iter()
        .chain(slice.puts.iter())
        .next()
        .map(|(strike, data)| SelectedOption {
            instrument_id: data.quote.instrument_id,
            strike: *strike,
            quote: data.quote,
            delta: data.greeks.as_ref().map(|greeks| greeks.delta),
        })
}

fn starting_balance(settlement_currency: Ustr) -> String {
    match settlement_currency.as_str() {
        "BTC" | "ETH" => format!("10 {settlement_currency}"),
        _ => format!("1000000 {settlement_currency}"),
    }
}

#[cfg(test)]
mod tests {
    use nautilus_core::UnixNanos;
    use rstest::rstest;

    use super::*;

    const NEAREST_EXPIRY_NS: u64 = 1_704_067_200_000_000_000;
    const LATER_EXPIRY_NS: u64 = 1_706_745_600_000_000_000;

    fn metadata(
        raw_symbol: &str,
        underlying: &str,
        settlement_currency: &str,
        expiration_ns: u64,
        strike: &str,
    ) -> OptionMetadata {
        OptionMetadata {
            instrument_id: InstrumentId::from(format!("{raw_symbol}.DERIBIT").as_str()),
            underlying: Ustr::from(underlying),
            settlement_currency: Ustr::from(settlement_currency),
            expiration_ns,
            strike: Price::from(strike),
        }
    }

    fn mixed_series_options() -> Vec<OptionMetadata> {
        vec![
            metadata(
                "BTC-20240101-45000-C",
                "BTC",
                "BTC",
                NEAREST_EXPIRY_NS,
                "45000",
            ),
            metadata(
                "BTC-20240101-50000-P",
                "BTC",
                "BTC",
                NEAREST_EXPIRY_NS,
                "50000",
            ),
            metadata("BTC-20240101-1-C", "BTC", "USDC", NEAREST_EXPIRY_NS, "1"),
            metadata("BTC-20240201-2-C", "BTC", "BTC", LATER_EXPIRY_NS, "2"),
        ]
    }

    #[rstest]
    fn test_nearest_series_filters_to_nearest_expiry_and_settlement_currency() {
        let options = mixed_series_options();

        let (series_id, instrument_ids, series_options) =
            nearest_series(Venue::new(VENUE), &options).unwrap();

        assert_eq!(series_id.venue, Venue::new(VENUE));
        assert_eq!(series_id.underlying, Ustr::from("BTC"));
        assert_eq!(series_id.settlement_currency, Ustr::from("BTC"));
        assert_eq!(series_id.expiration_ns, UnixNanos::from(NEAREST_EXPIRY_NS));
        assert_eq!(
            instrument_ids,
            vec![
                InstrumentId::from("BTC-20240101-45000-C.DERIBIT"),
                InstrumentId::from("BTC-20240101-50000-P.DERIBIT"),
            ]
        );
        assert_eq!(series_options.len(), 2);
        assert!(series_options.iter().all(|metadata| {
            metadata.expiration_ns == NEAREST_EXPIRY_NS
                && metadata.settlement_currency == Ustr::from("BTC")
        }));
    }

    #[rstest]
    fn test_default_strike_uses_selected_series_options_only() {
        let options = mixed_series_options();
        let (_series_id, _instrument_ids, series_options) =
            nearest_series(Venue::new(VENUE), &options).unwrap();

        let strike = median_strike(&series_options).unwrap();

        assert_eq!(strike, Price::from("50000"));
        assert!(
            !series_options
                .iter()
                .any(|metadata| metadata.strike == Price::from("1"))
        );
        assert!(
            !series_options
                .iter()
                .any(|metadata| metadata.strike == Price::from("2"))
        );
    }

    #[rstest]
    fn test_strike_range_preserves_delta_and_fixed_modes() {
        assert_eq!(
            strike_range(SelectionMode::Delta {
                target: 0.25,
                tolerance: 0.05,
            }),
            StrikeRange::Delta {
                target: 0.25,
                tolerance: 0.05,
            }
        );
        assert_eq!(
            strike_range(SelectionMode::Strike {
                strike: Price::from("50000"),
            }),
            StrikeRange::Fixed(vec![Price::from("50000")])
        );
    }
}
